### PR TITLE
Test helpers

### PR DESCRIPTION
**Pull request by maelle**

### CODE REVIEW

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+To summarize,
+
+| File                           | Run before tests | Loaded via `load_all()` | Installed with the package[^installed] | Testable[^testable] |
+|--------------------------------|------------------|-------------------------|----------------------------|----------|
+| tests/testthat/setup-.\*.R     | ✔️               | \-                      | \-                         | \-       |
+| tests/testthat/helper-.\*.R    | ✔️               | ✔️                      | \-                         | \-       |
+| R/any-name.R                   | ✔️               | ✔️                      | ✔️                         | ✔️       |
+| tests/testthat/anything-else.R | \-               | \-                      | \-                         | \-       |
```
Comment (gaborcsardi) [inline content/post/2020-11-18-testthat-utility-belt/index.Rmd:58]: I don't think you need the dashes for setup and helper files, so it is just `setup*.R`. Btw. is this a regexp? Then you need to escape the last dot. But it think it is easier to use globs for filenames, like the shells, i.e. `setup*.R`.

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+[^testable]: Yes, you could [test the code supporting your tests](https://github.com/r-lib/testthat/issues/1206#issuecomment-713583205).
+
+In practice,
+
+* In `tests/testthat/setup.R` you might do something like loading a package that helps your unit testing like `{vcr}`, `{httptest}` or `{presser}` if you're [testing an API client](https://books.ropensci.org/http-testing/).
+* In a helper like `tests/testthat/helper.R` or `R/test-helpers.R` you might define variables and functions that you'll use throughout your tests, even [custom skippers](https://testthat.r-lib.org/articles/skipping.html#helpers). To choose between the two locations, refer to the table above and your own needs and preferences. Note that if someone wanted to study testthat "utility belts" à la [Bob Rudis](https://rud.is/b/2018/04/08/dissecting-r-package-utility-belts/), they would probably only identify helper files like `tests/testthat/helper.R`.
+
+You'll notice testthat no longer recommends having a file with code to be run after tests... So how do you clean up after tests?
```
Comment (gaborcsardi) [inline content/post/2020-11-18-testthat-utility-belt/index.Rmd:70]: Can we just say teardown files (as well)?

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+Well, with `withr`'s various helper functions for deferring clean-up.
```
Comment (gaborcsardi) [inline content/post/2020-11-18-testthat-utility-belt/index.Rmd:71]: > Well, with `withr`'s various helper functions for deferring clean-up.

I think we don't use backticks for packages? Also. with -> use ?

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+So basically it means the code for cleaning lives near the code for making a mess.
+To learn more about this, read [the "self-cleaning text fixtures" vignette in testthat](https://testthat.r-lib.org/articles/test-fixtures.html).
```
Comment (gaborcsardi) [inline content/post/2020-11-18-testthat-utility-belt/index.Rmd:73]: Should we have an example and also mention `teardown_env()`?

**Review by gaborcsardi — APPROVED** [review]
Very cool!

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+## Code called in your tests
+
+Remember our post about [internal functions in R packages](/2019/12/12/internal-functions/)?
+What about internal functions in _unit tests_ of R packages?
+And code that needs to be run for tests?
+
+Where to put your code and function depends on where you'll want to use them.
+
+* It's best not to touch `tests/testthat/test.R`.
```
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:47]: oops I meant tests/testthat.R

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+* R scripts under `tests/testthat/` whose name start with `setup-` are loaded before tests are run but not with `devtools::load_all()` which means the code in there is _not_ available when you're interactively debugging a test. And yes, you'll be interactively debugging tests more often than you wish. :wink:
```
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:48]: can you expand on developing the tests? :thinking:

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd (hunk 2 above)
Comment (maelle) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:70]: well one should no longer use them?

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:48]: I mean, when you are writing the test cases. As opposed to debugging them.

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd (hunk 2 above)
Comment (gaborcsardi) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:70]: Right. Never mind then.

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd (hunk 6 above)
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:48]: When writing tests you'd use helper* then when you are done put the code in setup*, and when debugging you would add some browser() etc?

[the remark this answers is not in the document]
Comment (gaborcsardi) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:48]: I don't think I would move things around. For writing the tests youd' just need to source/run the setup part manually I suppose.

File: content/post/2020-11-18-testthat-utility-belt/index.Rmd (hunk 4 above)
Comment (maelle) [reply content/post/2020-11-18-testthat-utility-belt/index.Rmd:78]: I think the vignette is the best resource so won't add any example.

**Pull request merged**